oracle/oraclerun.go

```diff
@@ -0,0 +1,225 @@
+	logrus.AddHook(logmon.NewPrometheusHook())
+
+	prometheus.MustRegister(versionTotal)
+}
+
+// gatewayForwardedHeaders are HTTP headers which the grpc-gateway will encode
+// as grpc request metadata and forward to the oracle grpc server.  Forwarded
+// headers may be used for authentication flows, request tracing, etc.
+func (orc *Oracle) gatewayForwardedHeaders() []string {
+	return []string{
+		"Cookie",
+		"X-Forwarded-For",
+		"User-Agent",
+		"X-Forwarded-User-Agent",
+		"Referer",
+		orc.requestIDHeader,
+	}
+}
+
+func (orc *Oracle) incomingHeaderMatcher(h string) (string, bool) {
+	headers := orc.gatewayForwardedHeaders()
+
+	for i := range headers {
+		if strings.EqualFold(h, headers[i]) {
+			return h, true
+		}
+	}
+	return "", false
+}
+
+func (orc *Oracle) grpcGatewayMux() *runtime.ServeMux {
+	opts := []runtime.ServeMuxOption{
+		runtime.WithErrorHandler(svcerr.ErrIntercept(orc.log)),
+		runtime.WithIncomingHeaderMatcher(orc.incomingHeaderMatcher),
+		runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{
+			MarshalOptions: protojson.MarshalOptions{
+				UseProtoNames: true,
+			},
+			UnmarshalOptions: protojson.UnmarshalOptions{
+				DiscardUnknown: false,
+			},
+		}),
+	}
+
+	return runtime.NewServeMux(opts...)
+}
+
+func (orc *Oracle) grpcGateway(swaggerHandler http.Handler) (*runtime.ServeMux, http.Handler) {
+	jsonapi := orc.grpcGatewayMux()
+	pathOverides := midware.PathOverrides{
+		healthCheckPath: orc.healthCheckHandler(),
+	}
+	if swaggerHandler != nil {
+		pathOverides[swaggerPath] = swaggerHandler
+	}
+	middleware := midware.Chain{
+		// The trace header middleware appears early in the chain
+		// because of how important it is that they happen for essentially all
+		// requests.
+		midware.TraceHeaders(orc.requestIDHeader, true),
+		orc.addServerHeader(),
+		// PathOverrides and other middleware that may serve requests or have
+		// potential failure states should appear below here so they may rely
+		// on the presence of the generic utility middleware above.
+		pathOverides,
+	}
+
+	return jsonapi, middleware.Wrap(jsonapi)
+}
+
+type PortalConfig interface {
+	// RegisterServiceServer is required to be overidden by the implementation.
+	RegisterServiceServer(grpcServer *grpc.Server)
+	// RegisterServiceClient is required to be overidden by the implementation.
+	RegisterServiceClient(ctx context.Context, grpcCon *grpc.ClientConn, mux *runtime.ServeMux) error
+}
+
+func (orc *Oracle) Run(portalConfig PortalConfig) error {
+	orc.stateMut.Lock()
+	defer orc.stateMut.Unlock()
+	if orc.state != oracleStateInit {
+		return fmt.Errorf("run: invalid oracle state: %d", orc.state)
+	}
+	orc.state = oracleStateStarted
+
+	trySendError := func(c chan<- error, err error) {
+		if err == nil {
+			return
+		}
+		select {
+		case c <- err:
+		default:
+		}
+	}
+	errServe := make(chan error, 1)
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
Should `Run` really take a context that can be canceled to terminate the service?

Reply:
It seems reasonable to me, and is along the lines of how the `kong` library works:

https://github.com/alecthomas/kong